lib/mode.coffee

```diff
@@ -0,0 +1,234 @@
+  return messages
+
+constructMessage = (type, element) ->
+  message =
+    type: type
+    text: element.message
+    filePath: element.file
+    range: element.range
+  # children exists only in JSON messages
+  if element.children
+    message.trace = []
+    for children in element.children
+      message.trace.push
+        type: "Trace"
+        text: children.message
+        filePath: element.file
+        range: children.range or element.range
+  message
+
+buildRustcArguments = (linter, paths) ->
+  [editingFile, cargoManifestPath] = paths
+  Promise.resolve().then () =>
+    rustcArgs = switch linter.rustcBuildTest
+      when true then ['--cfg', 'test', '-Z', 'no-trans', '--color', 'never']
+      else ['-Z', 'no-trans', '--color', 'never']
+    cmd = [linter.rustcPath]
+      .concat rustcArgs
+    if cargoManifestPath
+      cmd.push '-L'
+      cmd.push path.join path.dirname(cargoManifestPath), linter.cargoDependencyDir
+    compilationFeatures = linter.compilationFeatures(false)
+    cmd = cmd.concat compilationFeatures if compilationFeatures
+    cmd = cmd.concat [editingFile]
+    [editingFile, cmd]
+
+buildCargoArguments = (linter, cargoManifestPath) ->
+  cargoArgs = switch linter.cargoCommand
+    when 'check' then ['check']
+    when 'test' then ['test', '--no-run']
+    when 'rustc' then ['rustc', '-Zno-trans', '--color', 'never']
+    when 'clippy' then ['clippy']
+    else ['build']
+
+  compilationFeatures = linter.compilationFeatures(true)
+  buildCargoPath(linter.cargoPath).then (cmd) ->
+    cmd = cmd
+      .concat cargoArgs
+      .concat ['-j', linter.jobsNumber]
+    cmd = cmd.concat compilationFeatures if compilationFeatures
+    cmd = cmd.concat ['--manifest-path', cargoManifestPath]
+    [cargoManifestPath, cmd]
+
+buildCargoPath = (cargoPath, cargoCommand) ->
+  usingMultitoolForClippy().then (canUseMultirust) ->
+    if cargoCommand == 'clippy' and canUseMultirust.result
+      [canUseMultirust.tool, 'run', 'nightly', 'cargo']
+    else
+      [cargoPath]
+
+usingMultitoolForClippy = () ->
+  # Try to use rustup
+  sb_exec.exec 'rustup', ['--version'], {ignoreExitCode: true}
+    .then ->
+      result: true, tool: 'rustup'
+    .catch ->
+      # Try to use odler multirust at least
+      sb_exec.exec 'multirust', ['--version'], {ignoreExitCode: true}
+        .then ->
+          result: true, tool: 'multirust'
+        .catch ->
+          result: false
+
+# These define the behabiour of each error mode linter-rust has
+errorModes =
+  JSON_RUSTC:
+    neededOutput: (stdout, stderr) ->
+      stderr
+
+    parse: (output, disabledWarnings) =>
+      parseJsonOutput output, disabledWarnings
+
+    buildArguments: (linter, file) ->
+      buildRustcArguments(linter, file).then (cmd_res) ->
+        [file, cmd] = cmd_res
+        cmd = cmd.concat ['--error-format=json']
+        [file, cmd]
+
+  JSON_CARGO:
+    neededOutput: (stdout, stderr) ->
+      stdout
+
+    parse: (output, disabledWarnings) ->
```

Comment:
🐛 Can the functionality here be merged into `parseJsonOutput`? They are _almost_ identical.